src/safe_provider.rs

```diff
@@ -0,0 +1,344 @@
+//!         ClientBuilder::default().ws(WsConnect::new("wss://localhost:8000")).await?,
+//!     );
+//!     let safe_provider = SafeProvider::new(provider)
+//!         .with_max_timeout(Duration::from_secs(30))
+//!         .with_max_retries(5);
+//!
+//!     let block = safe_provider.get_block_by_number(12345.into()).await?;
+//!     Ok(())
+//! }
+//! ```
+
+use std::{future::Future, time::Duration};
+
+use alloy::{
+    eips::BlockNumberOrTag,
+    network::Network,
+    providers::{Provider, RootProvider},
+    pubsub::Subscription,
+    rpc::types::{Filter, Log},
+    transports::{RpcError, TransportErrorKind},
+};
+use backon::{ExponentialBuilder, Retryable};
+use tracing::{debug, error};
+
+// RPC retry and timeout settings
+/// Default timeout used by `SafeProvider`
+pub const DEFAULT_MAX_TIMEOUT: Duration = Duration::from_secs(30);
+/// Default maximum number of retry attempts.
+pub const DEFAULT_MAX_RETRIES: usize = 5;
+/// Default base delay between retries.
+pub const DEFAULT_RETRY_INTERVAL: Duration = Duration::from_secs(1);
+
+/// Provider wrapper adding retries and timeouts.
+#[derive(Clone)]
+pub struct SafeProvider<N: Network> {
+    provider: RootProvider<N>,
+    max_timeout: Duration,
+    max_retries: usize,
+    retry_interval: Duration,
+}
+
+impl<N: Network> SafeProvider<N> {
+    /// Create a new `SafeProvider` with default settings.
+    #[must_use]
+    pub fn new(provider: RootProvider<N>) -> Self {
+        Self {
+            provider,
+            max_timeout: DEFAULT_MAX_TIMEOUT,
+            max_retries: DEFAULT_MAX_RETRIES,
+            retry_interval: DEFAULT_RETRY_INTERVAL,
+        }
+    }
+
+    #[must_use]
+    pub fn with_max_timeout(mut self, timeout: Duration) -> Self {
+        self.max_timeout = timeout;
+        self
+    }
+
+    #[must_use]
+    pub fn with_max_retries(mut self, max_retries: usize) -> Self {
+        self.max_retries = max_retries;
+        self
+    }
+
+    #[must_use]
+    pub fn with_retry_interval(mut self, retry_interval: Duration) -> Self {
+        self.retry_interval = retry_interval;
+        self
+    }
+
+    #[must_use]
+    pub fn inner(&self) -> &RootProvider<N> {
+        &self.provider
+    }
+
+    /// Fetch a block by number with retry and timeout.
+    ///
+    /// # Errors
+    /// Returns `RpcError<TransportErrorKind>` if the RPC call fails
+    /// after exhausting retries or times out.
+    pub async fn get_block_by_number(
+        &self,
+        number: BlockNumberOrTag,
+    ) -> Result<Option<N::BlockResponse>, RpcError<TransportErrorKind>> {
+        debug!("SafeProvider eth_getBlockByNumber called with number: {:?}", number);
+        let provider = self.provider.clone();
+        let result = self
+            .retry_with_total_timeout(|| async { provider.get_block_by_number(number).await })
+            .await;
+        if let Err(e) = &result {
+            error!("SafeProvider eth_getByBlockNumber failed: {}", e);
```

Comment:
```suggestion
            error!(error = e, "SafeProvider eth_getByBlockNumber failed");
```
Also, maybe no need to mention the underlying `SafeProvider` type in logs? Just a notification that an operation failed should be sufficient

Reply:
You cant do the error = e because error needs to implement some like tracing trait thing 

Yeah i added it because i wasnt sure if it was obvious where the error came from if it was just like eth_getByBlockNumber failed..

Removed though perhaps its useless :)